Use library for memleak checks

rikardfalkeborn opened the pull request:
Change `memleakOnRealloc` and `leakReturnValNotUsed` to use library
configuration instead of hardcoding "`realloc`".

In order to do so, some care needs to be taken when matching for a
reallocation function, since it can no longer be assumed that the input
to be allocated is the first argument of the function. This complicates
`getReallocationType()` and `checkReallocUsage()` but is necessary in order
to handle for example `freopen()` properly.

Also, refactor `memleakOnRealloc` check to reduce duplicated code when
checking "`a`" and "`*a`". When doing so, extending the check to look for
arbitrary number of "*" can be done for free (just change an if
statement to a while statement). Most likely, this is an unusual case in
real world code.

rikardfalkeborn commented:
> looks good. let's see what CI says. maybe I need to restart it.

Appveyor failed, but it seems unrelated to my changes.

rikardfalkeborn commented:
Appveyor seems really slow at the moment. When things stabilise, I can merge with master and amend the PR.

danmar commented on lib/checkmemoryleak.cpp:
```diff
@@ -177,12 +178,26 @@ CheckMemoryLeak::AllocType CheckMemoryLeak::getReallocationType(const Token *tok
+    if (!(f && f->reallocArg > 0 && f->reallocArg <= numberOfArguments(tok2)))
+        return No;
+    const Token* arg = getArguments(tok2).at(f->reallocArg - 1);
+    while (Token::simpleMatch(arg, "*"))
```
I would prefer:

    while (arg && arg->isUnaryOp("*"))
        arg = arg->astOperand1();

That should for instance handle such code also: `*(p)`

rikardfalkeborn replied:
Thanks, I'll update.

danmar commented on lib/checkmemoryleak.cpp:
```diff
@@ -488,16 +500,35 @@ void CheckMemoryLeakInFunction::checkReallocUsage()
-                isNoArgument(symbolDatabase, tok->varId())) {
+            if (tok->varId() > 0 && Token::Match(tok, "%name% =")) {
+                const Token* reallocTok = tok->tokAt(2);
+                if (reallocTok && reallocTok->str() == "(") {
```
This code looks fragile.. I think `Token::isCast` method can be used here to make the code more robust. Please consider such code:

    p = (realloc(p,100));

.. it's probably better to use the AST to look at assignment RHS.

rikardfalkeborn replied:
I tried that, but didn't get it to work so I more or less copied how casts were treated in other places in the file, but I now realize that it's because the checks are run on the full token list, but the tests are still run on the simplified token list. I think this also explains why I never managed to write a test case that failed when removing these casts, but I needed them there to get `make checkcfg` working...

I'll update it.

rikardfalkeborn replied:
Hmm, as far as I can see, no checks run on the simplified token list anymore, but quite a few testcases run `simplifyTokenList2()`. This must be incorrect?

danmar replied:
> Hmm, as far as I can see, no checks run on the simplified token list anymore, but quite a few testcases run simplifyTokenList2(). This must be incorrect?

yes that is very unfortunate.

rikardfalkeborn commented:
> thanks for working on this. However this is old code that was written before we had AST... it's not very well written.

Is there a reason for still keeping two separate files `checkmemoryleak.cpp` and `checkleakautovar.cpp`? It seems they check more or less the same things?

danmar commented:
> Is there a reason for still keeping two separate files checkmemoryleak.cpp and checkleakautovar.cpp? It seems they check more or less the same things?

No there is no reason. If we can remove some checking in checkmemoryleak.cpp without getting too much false negatives, please remove it.

rikardfalkeborn commented:
I switched to use the full tokenlist in the memleakOnRealloc tests (all should be switched but I figured this PR was big enough). It revealed a couple of false negatives and false positives. The false negatives are fixed in this PR (they were related to casts which were not handled). Unfortunately, I was not able to fix the false positives (I marked them as TODOs for now). The false positives look like:
```
void foo()
{
   char *a = 0;
    if ((a = realloc(a, 100)) == NULL)
       return;
   free(a);
}
```
I was not able to fix this (I tried to use `reallocTok->getValues(0)`, but it always returned `nullptr`). Is this a limitation in `ValueFlow` or am I using it wrong?

Note that the FP is not introduced by this PR, it's only shown since I switched to the full tokenlist.

danmar commented:
> I was not able to fix this (I tried to use reallocTok->getValues(0), but it always returned nullptr). Is this a limitation in ValueFlow or am I using it wrong?

strange. There is no such limitation in ValueFlow.

danmar commented:
> Note that the FP is not introduced by this PR, it's only shown since I switched to the full tokenlist.

ok.. then do not worry about that now.

danmar commented on lib/checkmemoryleak.cpp:
```diff
@@ -488,16 +502,46 @@ void CheckMemoryLeakInFunction::checkReallocUsage()
+                if (!((allocType == Malloc || allocType == OtherMem)))
+                    continue;
+                const Token* arg = getArguments(reallocTok).at(f->reallocArg - 1);
+                while (arg && arg->isCast())
```
we should consider `static_cast` etc also.. but I leave that for now.

rikardfalkeborn commented:
Ok, so after some digging, I've concluded that it is a `ValueFlow` issue:
```
void f() {
    void * a;
    void * b;
    void * c;
    a = realloc(a,10); // valueflow does not know the value about a passed as argument here
    c = realloc(b,10); // valueflow knows b is zero here
}
```
Note however that there are no FPs when `NULL` is used instead `0` since when checking if the pointer is saved in another pointer before, one matching condition is `%varid% = %name%` where `NULL` seems to match, which leads to these FPs are probably not very common. However, the way the condition is written, there will be no warning for the following:
```
void f() {
    void *a = NULL;
    a = malloc(10);
    a = realloc(a, 20); // No warning here due to a = NULL above
}
``` 
I'll create a trac tickets about the valueflow issue and the FN above.

danmar commented:
ok thanks.. nice work!
